### ISSUE TITLE
7.5" V2 display driver does not work

### ISSUE THREAD

**Issue by olback** [caemor/epd-waveshare#70]
A few things I found while trying to get my display to work as expected:
* `epd7in5_v2v::EPD7in5::new(...)` hangs [here](https://github.com/caemor/epd-waveshare/blob/main/src/epd7in5_v2/mod.rs#L72) since the `wait_until_idle` implementation is wrong. In the [example C code provided by Waveshare](https://github.com/waveshare/e-Paper/blob/7c368a22afb5456fb4622c617e792bd2f83e878b/RaspberryPi_JetsonNano/c/lib/e-Paper/EPD_7in5_V2.c#L78) one can see that the driver sends a `GET_STATUS` command and waits 20ms.  This seems to be needed otherwise the busy pin wont update which is the root cause to the problem.

* Black and White seems to have been mixed up. Setting the fill-color to black makes the display white and vice versa. Same thing goes for text colors/background.

**A few things that might be important:**
* Cross-compiling from Arch Linux x64_64 to aarch64 with `aarch64-linux-gnu-gcc` targeting `aarch64-unknown-linux-musl`.
* Using the [rppal](https://github.com/golemparts/rppal) crate with the hal-unproven feature.
* Running on Raspberry Pi OS 64-bit Beta (Kernel 5.4.51-v8+) on a Raspberry Pi 3b+.

A hacked together version of epd-waveshare can be found [here](https://github.com/olback/epd-waveshare/tree/7.5hack) and an example can be found [here](https://github.com/olback/smart-home/blob/epd-working/portal/src/main.rs).

**Comment by caemor** [comment]
Merged #71

**Comment by caemor** [comment]
What is still missing from here:
- black/white stuff  /cc @whiite 
- Optionally a testcase for the delay

**Comment by Flocksserver** [comment]
Hey :) 
I am facing the same problem. My proof of concept hangs at `epd7in5_v2v::EPD7in5::new`. I am not so familiar with the hardware related programming and the c code provided by waveshare. Is there any plan to fix this? Or do we have to find a proper fix ourselves and create a PR? 
Thanks for your work 💪

**Comment by caemor** [comment]
Hello Flocksserver,
the delay issue was already fixed and merged in #71 
So your issue is most likely slightly different. Which version of this crate are you using? There was quite a long break of crates.io releases from my end until this week, so you might still be stuck on the old, unpatched version.
Another possible failure could be your display. Are you sure you have a v2 or maybe you got the newest v3 that they are slowly rolling out for the various displays? Or maybe we got a regression or bug somewhere else

**Comment by Flocksserver** [comment]
Wow your second name is "The Flash". That was a fast reply :) 
I got it running, but I don't know what exactly the issue was. Probably not in this crate. 
The change was, that now I am using the rppal dependency (olback mentioned this crate) for initializing spi, pins etc. I use an old RPI 1 Model B with armv6. Maybe there is an issue in linux_embedded_hal for that architecture, who knows. Not your problem 😎 🥳

**Comment by whiite** [comment]
Got a chance recently to take a look at this and I believe the culprit might be the change made in #83 which writes bytes one at time instead of all at once as reverting this got my 7.5" HD display working again and suspect this may be the case for the V2 too.
Happy to make a PR to sort this one out but seems like we may need to support both depending on the display, don't suppose we know which screens will need each write method?

**Comment by GBouerat** [comment]
Hi,

I just bought a "7.5 Inch B/W/R V2/V3" with the e-Paper Driver HAT https://www.waveshare.com/7.5inch-e-Paper-HAT-B.htm

And after a day, I didn't succeed to display something on the screen with a raspberry 4 and a raspberry 3 😞
(the c demo from the documentation works https://www.waveshare.com/wiki/7.5inch_e-Paper_HAT_(B)_Manual#C_2)

Can someone help me or provide a working sample ? please,

Thanks

**Comment by whiite** [comment]
Hey, out of curiosity does this fork work for you instead: https://github.com/whiite/epd-waveshare
This includes the change I talked about above that was needed for my 7.5” HD to work

**Comment by caemor** [comment]
@whiite If you have time, could you test #141 (branch fix_single_write_for_7in5)? I reverted the change of #83 for all 7in5 displays by adding a const switch case.

**Comment by whiite** [comment]
@caemor yeah sure thing, will probably be next week that I get the chance and I can only check for the HD version at the moment. Looks like a neat solution!

**Comment by GBouerat** [comment]
> @GBouerat out of curiosity does this fork work for you instead: https://github.com/whiite/epd-waveshare This includes the change I talked about above that was needed for my 7.5” HD to work

I try everything I found include this fork, and nothing works for me so far, that's why I ask for some help 👍

**Comment by caemor** [comment]
Since I hope this got solved by #141 I'm going to close this for now. If it is still an issue, feel free to either write a comment or open a new issue with for the specific issue.

**Comment by dexterdy** [comment]
Hey there, commenting because I have the same issue. I am using the raspberry pi zero 2w.

I first ran into another issue when running the example, which is that the exporting the cs pin causes os errors. When running gpioinfo on the rpi, it seems that by enabling spi using raspi-config, the cs pins are already active and can't be exported again:

```
$ gpioinfo
gpiochip0 - 54 lines:
        line   0:       "ID_SDA"                input
        line   1:       "ID_SCL"                input
        line   2:       "GPIO2"                 input
        line   3:       "GPIO3"                 input
        line   4:       "GPIO4"                 input
        line   5:       "GPIO5"                 input
        line   6:       "GPIO6"                 input
        line   7:       "GPIO7"                 output active-low consumer="spi0 CS1"
        line   8:       "GPIO8"                 output active-low consumer="spi0 CS0"
        ...
```

I removed the block of code that tries to manipulate the cs pin. 

Now I am running into the exact same issue, which is that the code seems to be waiting indefinitely for the display to be ready.

edit: I have tried the rppal library, which didn't work either.
